File: src/analyzer.ts
/**
 * @file analyzer.ts
 * @module atomis/analyzer
 *
 * Semantic analyzer for the Atomis AST. It walks a {@link Program} and
 * performs:
 *
 *  - Scope resolution — tracks declared names per lexical scope.
 *  - Atom tracking — records every `atom` for reactivity wiring.
 *  - Basic type checking — flags a few obvious mismatches.
 *  - GhostNet validation — unique node/channel/mesh ids; `connect` statements
 *    must reference declared channels and nodes; transports must be one of
 *    `bluetooth | tcp | ws | ble5`.
 *  - Purity checking — warns when a `pure fn` performs obviously impure work
 *    (await, console/output, assignment to an outer atom, GhostNet calls).
 *
 * Results are returned as a structured {@link AnalysisResult} containing a flat
 * list of {@link Diagnostic} objects plus the collected symbol metadata used by
 * later stages.
 */

import {
  AtomDecl,
  ChannelDecl,
  ConnectStmt,
  FnDecl,
  MeshDecl,
  NodeDecl,
  Program,
  PureFnDecl,
  Statement,
  TSPassthrough,
} from "./ast";

/** Severity of a diagnostic. */
export type Severity = "error" | "warn";

/** A single structured diagnostic message. */
export interface Diagnostic {
  /** Human-readable message. */
  message: string;
  /** 1-based source line. */
  line: number;
  /** 1-based source column. */
  col: number;
  /** Diagnostic severity. */
  severity: Severity;
}

/** Recorded metadata about a declared `atom`. */
export interface AtomInfo {
  name: string;
  typeAnnotation?: string;
  line: number;
}

/** The result of analyzing a program. */
export interface AnalysisResult {
  /** All diagnostics, in source order. */
  diagnostics: Diagnostic[];
  /** Atoms discovered (for reactivity wiring). */
  atoms: AtomInfo[];
  /** Declared GhostNet node ids. */
  nodeIds: string[];
  /** Declared GhostNet channel ids. */
  channelIds: string[];
  /** Declared GhostNet mesh ids. */
  meshIds: string[];
  /** Whether any error-severity diagnostic was produced. */
  hasErrors: boolean;
}

/** Valid GhostNet transports. */
const VALID_TRANSPORTS = new Set(["bluetooth", "tcp", "ws", "ble5"]);

/** A lexical scope mapping declared names to the line they were declared on. */
interface Scope {
  names: Map<string, number>;
  parent?: Scope;
}

/**
 * The Atomis semantic analyzer.
 */
export class Analyzer {
  private readonly diagnostics: Diagnostic[] = [];
  private readonly atoms: AtomInfo[] = [];
  private readonly nodeIds = new Map<string, number>();
  private readonly channelIds = new Map<string, number>();
  private readonly meshIds = new Map<string, number>();
  private scope: Scope = { names: new Map() };

  /**
   * Analyze a program.
   * @param program The parsed {@link Program} AST.
   * @returns Structured {@link AnalysisResult}.
   */
  public analyze(program: Program): AnalysisResult {
    this.visitBlock(program.body);
    this.validateConnects(program);

    return {
      diagnostics: this.diagnostics,
      atoms: this.atoms,
      nodeIds: [...this.nodeIds.keys()],
      channelIds: [...this.channelIds.keys()],
      meshIds: [...this.meshIds.keys()],
      hasErrors: this.diagnostics.some((d) => d.severity === "error"),
    };
  }

  /* ── diagnostics helpers ───────────────────────────────────────────── */

  private error(message: string, line: number, col: number): void {
    this.diagnostics.push({ message, line, col, severity: "error" });
  }

  private warn(message: string, line: number, col: number): void {
    this.diagnostics.push({ message, line, col, severity: "warn" });
  }

  /* ── scope helpers ─────────────────────────────────────────────────── */

  private pushScope(): void {
    this.scope = { names: new Map(), parent: this.scope };
  }

  private popScope(): void {
    if (this.scope.parent) this.scope = this.scope.parent;
  }

  private declare(name: string, line: number, col: number): void {
    if (this.scope.names.has(name)) {
      this.warn(`'${name}' is already declared in this scope`, line, col);
    }
    this.scope.names.set(name, line);
  }

  /* ── traversal ─────────────────────────────────────────────────────── */

  /**
   * Visit a list of statements within the current scope.
   * @param body Statements to visit.
   */
  private visitBlock(body: Statement[]): void {
    // First pass: register declarations so order doesn't cause false positives.
    for (const stmt of body) this.preregister(stmt);
    // Second pass: validate each statement.
    for (const stmt of body) this.visit(stmt);
  }

  private preregister(stmt: Statement): void {
    switch (stmt.type) {
      case "AtomDecl":
        this.declare(stmt.name, stmt.line, stmt.col);
        break;
      case "FnDecl":
      case "PureFnDecl":
        this.declare(stmt.name, stmt.line, stmt.col);
        break;
      default:
        break;
    }
  }

  private visit(stmt: Statement): void {
    switch (stmt.type) {
      case "AtomDecl":
        this.visitAtom(stmt);
        break;
      case "FnDecl":
        this.visitFn(stmt, false);
        break;
      case "PureFnDecl":
        this.visitFn(stmt, true);
        break;
      case "NodeDecl":
        this.visitNode(stmt);
        break;
      case "ChannelDecl":
        this.visitChannel(stmt);
        break;
      case "MeshDecl":
        this.visitMesh(stmt);
        break;
      case "GuardStmt":
        this.pushScope();
        this.visitBlock(stmt.elseBody);
        this.popScope();
        break;
      case "CellBlock":
        this.pushScope();
        this.visitBlock(stmt.body);
        this.popScope();
        break;
      // ConnectStmt validated separately (needs full id sets first).
      // EncryptDecl, RenderBlock, ImportDecl, MetaBlock, OutputCall,
      // MatchExpr, TSPassthrough: no structural checks here.
      default:
        break;
    }
  }

  /* ── atom checks ───────────────────────────────────────────────────── */

  private visitAtom(stmt: AtomDecl): void {
    this.atoms.push({
      name: stmt.name,
      typeAnnotation: stmt.typeAnnotation,
      line: stmt.line,
    });
    this.checkBasicType(stmt);
  }

  /**
   * Very small "obvious mismatch" type check: when an atom has an explicit
   * primitive type and a literal initializer that clearly disagrees.
   */
  private checkBasicType(stmt: AtomDecl): void {
    const ty = stmt.typeAnnotation?.trim();
    const init = stmt.initializer?.trim();
    if (!ty || !init) return;

    const isStringLiteral = /^["'`]/.test(init);
    const isNumberLiteral = /^-?\d/.test(init);
    const isBoolLiteral = init === "true" || init === "false";
    const isArrayLiteral = init.startsWith("[");

    if (ty === "string" && (isNumberLiteral || isBoolLiteral)) {
      this.error(`Type mismatch: atom '${stmt.name}' is string but initialized with ${init}`, stmt.line, stmt.col);
    } else if (ty === "number" && (isStringLiteral || isBoolLiteral)) {
      this.error(`Type mismatch: atom '${stmt.name}' is number but initialized with ${init}`, stmt.line, stmt.col);
    } else if (ty === "boolean" && (isStringLiteral || isNumberLiteral)) {
      this.error(`Type mismatch: atom '${stmt.name}' is boolean but initialized with ${init}`, stmt.line, stmt.col);
    } else if (ty.endsWith("[]") && init !== "[]" && !isArrayLiteral && init !== "undefined") {
      this.warn(`atom '${stmt.name}' is typed as an array but initializer is not an array literal`, stmt.line, stmt.col);
    }
  }

  /* ── function checks ───────────────────────────────────────────────── */

  private visitFn(stmt: FnDecl | PureFnDecl, pure: boolean): void {
    this.pushScope();
    for (const p of stmt.params) this.declare(p.name, p.line, p.col);
    this.visitBlock(stmt.body);
    if (pure) this.checkPurity(stmt);
    this.popScope();
  }

  /**
   * Warn when a `pure fn` performs obviously impure operations.
   */
  private checkPurity(stmt: PureFnDecl | FnDecl): void {
    if (stmt.isAsync) {
      this.warn(`pure fn '${stmt.name}' is async; pure functions should not perform I/O`, stmt.line, stmt.col);
    }
    const impure = /\b(await|console|output|fetch|Math\.random|Date\.now)\b|\.send\(|\.scan\(/;
    for (const s of this.flattenBody(stmt.body)) {
      if (s.type === "OutputCall") {
        this.warn(`pure fn '${stmt.name}' calls output(); pure functions must not produce side effects`, s.line, s.col);
        continue;
      }
      const raw = (s as TSPassthrough).raw;
      if (typeof raw === "string" && impure.test(raw)) {
        this.warn(`pure fn '${stmt.name}' appears to perform a side effect: ${raw.trim().slice(0, 40)}`, s.line, s.col);
      }
    }
  }

  /** Flatten a body (recursing into guard/cell blocks) into a flat list. */
  private flattenBody(body: Statement[]): Statement[] {
    const out: Statement[] = [];
    for (const s of body) {
      out.push(s);
      if (s.type === "GuardStmt") out.push(...this.flattenBody(s.elseBody));
      else if (s.type === "CellBlock") out.push(...this.flattenBody(s.body));
    }
    return out;
  }

  /* ── GhostNet checks ───────────────────────────────────────────────── */

  private visitNode(stmt: NodeDecl): void {
    if (this.nodeIds.has(stmt.id)) {
      this.error(`Duplicate node id '${stmt.id}' (first declared on line ${this.nodeIds.get(stmt.id)})`, stmt.line, stmt.col);
    } else {
      this.nodeIds.set(stmt.id, stmt.line);
    }
  }

  private visitChannel(stmt: ChannelDecl): void {
    if (this.channelIds.has(stmt.id)) {
      this.error(`Duplicate channel id '${stmt.id}' (first declared on line ${this.channelIds.get(stmt.id)})`, stmt.line, stmt.col);
    } else {
      this.channelIds.set(stmt.id, stmt.line);
    }
  }

  private visitMesh(stmt: MeshDecl): void {
    if (this.meshIds.has(stmt.id)) {
      this.error(`Duplicate mesh id '${stmt.id}' (first declared on line ${this.meshIds.get(stmt.id)})`, stmt.line, stmt.col);
    } else {
      this.meshIds.set(stmt.id, stmt.line);
    }
  }

  /**
   * Validate `connect` statements after all declarations have been collected.
   * Recurses through nested blocks to find every connect statement.
   */
  private validateConnects(program: Program): void {
    const all = this.collectConnects(program.body);
    for (const c of all) {
      if (!this.channelIds.has(c.channel)) {
        this.error(`connect references undeclared channel '${c.channel}'`, c.line, c.col);
      }
      if (!this.nodeIds.has(c.node)) {
        this.error(`connect references undeclared node '${c.node}'`, c.line, c.col);
      }
      if (!VALID_TRANSPORTS.has(c.transport)) {
        this.error(
          `Invalid transport '${c.transport}'; expected one of bluetooth | tcp | ws | ble5`,
          c.line,
          c.col,
        );
      }
    }
  }

  private collectConnects(body: Statement[]): ConnectStmt[] {
    const out: ConnectStmt[] = [];
    for (const s of body) {
      if (s.type === "ConnectStmt") out.push(s);
      else if (s.type === "GuardStmt") out.push(...this.collectConnects(s.elseBody));
      else if (s.type === "CellBlock") out.push(...this.collectConnects(s.body));
      else if (s.type === "FnDecl" || s.type === "PureFnDecl") {
        out.push(...this.collectConnects(s.body));
      }
    }
    return out;
  }
}

/**
 * Convenience wrapper: analyze a program in one call.
 * @param program The parsed program.
 * @returns The analysis result.
 */
export function analyze(program: Program): AnalysisResult {
  return new Analyzer().analyze(program);
}

File: atomis-rs/src/analyzer.rs
//! Semantic analyzer for the Atomis AST.
//!
//! Faithful port of `src/analyzer.ts`. Walks a `Program` and produces a flat,
//! source-ordered list of `Diagnostic`s plus collected symbol metadata. It does
//! NOT transform. Diagnostic *order* matters (the CLI prints them in order), so
//! the traversal mirrors the TS reference exactly: per block, a preregister pass
//! (which can warn on duplicates) precedes the validate pass; `connect`
//! statements are validated last.

use crate::ast::*;
use crate::parser::contains_word;
use std::collections::HashMap;

/// Severity of a diagnostic.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Severity {
    Error,
    Warn,
}

/// A single structured diagnostic message.
#[derive(Debug, Clone)]
pub struct Diagnostic {
    pub message: String,
    pub line: usize,
    pub col: usize,
    pub severity: Severity,
}

/// Recorded metadata about a declared `atom`.
#[derive(Debug, Clone)]
pub struct AtomInfo {
    pub name: String,
    pub type_annotation: Option<String>,
    pub line: usize,
}

/// The result of analyzing a program.
#[derive(Debug, Clone)]
pub struct AnalysisResult {
    pub diagnostics: Vec<Diagnostic>,
    pub atoms: Vec<AtomInfo>,
    pub node_ids: Vec<String>,
    pub channel_ids: Vec<String>,
    pub mesh_ids: Vec<String>,
    pub has_errors: bool,
}

const VALID_TRANSPORTS: &[&str] = &["bluetooth", "tcp", "ws", "ble5"];

/// A lexical scope mapping declared names to the line they were declared on.
struct Scope {
    names: HashMap<String, usize>,
}

/// The Atomis semantic analyzer.
pub struct Analyzer {
    diagnostics: Vec<Diagnostic>,
    atoms: Vec<AtomInfo>,
    // Insertion-ordered id maps (value = first-declared line).
    node_ids: Vec<(String, usize)>,
    channel_ids: Vec<(String, usize)>,
    mesh_ids: Vec<(String, usize)>,
    scopes: Vec<Scope>,
}

impl Analyzer {
    pub fn new() -> Self {
        Analyzer {
            diagnostics: Vec::new(),
            atoms: Vec::new(),
            node_ids: Vec::new(),
            channel_ids: Vec::new(),
            mesh_ids: Vec::new(),
            scopes: vec![Scope { names: HashMap::new() }],
        }
    }

    pub fn analyze(mut self, program: &Program) -> AnalysisResult {
        self.visit_block(&program.body);
        self.validate_connects(program);

        let has_errors = self.diagnostics.iter().any(|d| d.severity == Severity::Error);
        AnalysisResult {
            diagnostics: self.diagnostics,
            atoms: self.atoms,
            node_ids: self.node_ids.into_iter().map(|(k, _)| k).collect(),
            channel_ids: self.channel_ids.into_iter().map(|(k, _)| k).collect(),
            mesh_ids: self.mesh_ids.into_iter().map(|(k, _)| k).collect(),
            has_errors,
        }
    }

    /* ── diagnostics helpers ───────────────────────────────────────────── */

    fn error(&mut self, message: String, line: usize, col: usize) {
        self.diagnostics.push(Diagnostic {
            message,
            line,
            col,
            severity: Severity::Error,
        });
    }

    fn warn(&mut self, message: String, line: usize, col: usize) {
        self.diagnostics.push(Diagnostic {
            message,
            line,
            col,
            severity: Severity::Warn,
        });
    }

    /* ── scope helpers ─────────────────────────────────────────────────── */

    fn push_scope(&mut self) {
        self.scopes.push(Scope { names: HashMap::new() });
    }

    fn pop_scope(&mut self) {
        if self.scopes.len() > 1 {
            self.scopes.pop();
        }
    }

    fn declare(&mut self, name: &str, line: usize, col: usize) {
        let exists = self.scopes.last().unwrap().names.contains_key(name);
        if exists {
            self.warn(
                format!("'{}' is already declared in this scope", name),
                line,
                col,
            );
        }
        self.scopes
            .last_mut()
            .unwrap()
            .names
            .insert(name.to_string(), line);
    }

    /* ── traversal ─────────────────────────────────────────────────────── */

    fn visit_block(&mut self, body: &[Statement]) {
        // First pass: register declarations so order doesn't cause false positives.
        for stmt in body {
            self.preregister(stmt);
        }
        // Second pass: validate each statement.
        for stmt in body {
            self.visit(stmt);
        }
    }

    fn preregister(&mut self, stmt: &Statement) {
        match stmt {
            Statement::AtomDecl(s) => self.declare(&s.name, s.line, s.col),
            Statement::FnDecl(s) | Statement::PureFnDecl(s) => {
                self.declare(&s.name, s.line, s.col)
            }
            _ => {}
        }
    }

    fn visit(&mut self, stmt: &Statement) {
        match stmt {
            Statement::AtomDecl(s) => self.visit_atom(s),
            Statement::FnDecl(s) => self.visit_fn(s, false),
            Statement::PureFnDecl(s) => self.visit_fn(s, true),
            Statement::NodeDecl(s) => self.visit_node(s),
            Statement::ChannelDecl(s) => self.visit_channel(s),
            Statement::MeshDecl(s) => self.visit_mesh(s),
            Statement::GuardStmt(s) => {
                self.push_scope();
                self.visit_block(&s.else_body);
                self.pop_scope();
            }
            Statement::CellBlock(s) => {
                self.push_scope();
                self.visit_block(&s.body);
                self.pop_scope();
            }
            _ => {}
        }
    }

    /* ── atom checks ───────────────────────────────────────────────────── */

    fn visit_atom(&mut self, stmt: &AtomDecl) {
        self.atoms.push(AtomInfo {
            name: stmt.name.clone(),
            type_annotation: stmt.type_annotation.clone(),
            line: stmt.line,
        });
        self.check_basic_type(stmt);
    }

    fn check_basic_type(&mut self, stmt: &AtomDecl) {
        let ty = match &stmt.type_annotation {
            Some(t) => t.trim().to_string(),
            None => return,
        };
        let init = match &stmt.initializer {
            Some(i) => i.trim().to_string(),
            None => return,
        };
        if ty.is_empty() || init.is_empty() {
            return;
        }

        let first = init.chars().next().unwrap_or('\0');
        let is_string_literal = first == '"' || first == '\'' || first == '`';
        let is_number_literal = {
            // /^-?\d/
            let mut cs = init.chars();
            let c0 = cs.next().unwrap_or('\0');
            if c0 == '-' {
                cs.next().map(|c| c.is_ascii_digit()).unwrap_or(false)
            } else {
                c0.is_ascii_digit()
            }
        };
        let is_bool_literal = init == "true" || init == "false";
        let is_array_literal = init.starts_with('[');

        if ty == "string" && (is_number_literal || is_bool_literal) {
            self.error(
                format!(
                    "Type mismatch: atom '{}' is string but initialized with {}",
                    stmt.name, init
                ),
                stmt.line,
                stmt.col,
            );
        } else if ty == "number" && (is_string_literal || is_bool_literal) {
            self.error(
                format!(
                    "Type mismatch: atom '{}' is number but initialized with {}",
                    stmt.name, init
                ),
                stmt.line,
                stmt.col,
            );
        } else if ty == "boolean" && (is_string_literal || is_number_literal) {
            self.error(
                format!(
                    "Type mismatch: atom '{}' is boolean but initialized with {}",
                    stmt.name, init
                ),
                stmt.line,
                stmt.col,
            );
        } else if ty.ends_with("[]") && init != "[]" && !is_array_literal && init != "undefined" {
            self.warn(
                format!(
                    "atom '{}' is typed as an array but initializer is not an array literal",
                    stmt.name
                ),
                stmt.line,
                stmt.col,
            );
        }
    }

    /* ── function checks ───────────────────────────────────────────────── */

    fn visit_fn(&mut self, stmt: &FnDeclData, pure: bool) {
        self.push_scope();
        for p in &stmt.params {
            self.declare(&p.name, p.line, p.col);
        }
        self.visit_block(&stmt.body);
        if pure {
            self.check_purity(stmt);
        }
        self.pop_scope();
    }

    fn check_purity(&mut self, stmt: &FnDeclData) {
        if stmt.is_async {
            self.warn(
                format!(
                    "pure fn '{}' is async; pure functions should not perform I/O",
                    stmt.name
                ),
                stmt.line,
                stmt.col,
            );
        }
        let flat = flatten_body(&stmt.body);
        for s in flat {
            if let Statement::OutputCall(o) = s {
                self.warn(
                    format!(
                        "pure fn '{}' calls output(); pure functions must not produce side effects",
                        stmt.name
                    ),
                    o.line,
                    o.col,
                );
                continue;
            }
            if let Statement::TSPassthrough(p) = s {
                if impure_test(&p.raw) {
                    let snippet: String = p.raw.trim().chars().take(40).collect();
                    self.warn(
                        format!(
                            "pure fn '{}' appears to perform a side effect: {}",
                            stmt.name, snippet
                        ),
                        p.line,
                        p.col,
                    );
                }
            }
        }
    }

    /* ── GhostNet checks ───────────────────────────────────────────────── */

    fn visit_node(&mut self, stmt: &NodeDecl) {
        if let Some(line) = find_id(&self.node_ids, &stmt.id) {
            self.error(
                format!(
                    "Duplicate node id '{}' (first declared on line {})",
                    stmt.id, line
                ),
                stmt.line,
                stmt.col,
            );
        } else {
            self.node_ids.push((stmt.id.clone(), stmt.line));
        }
    }

    fn visit_channel(&mut self, stmt: &ChannelDecl) {
        if let Some(line) = find_id(&self.channel_ids, &stmt.id) {
            self.error(
                format!(
                    "Duplicate channel id '{}' (first declared on line {})",
                    stmt.id, line
                ),
                stmt.line,
                stmt.col,
            );
        } else {
            self.channel_ids.push((stmt.id.clone(), stmt.line));
        }
    }

    fn visit_mesh(&mut self, stmt: &MeshDecl) {
        if let Some(line) = find_id(&self.mesh_ids, &stmt.id) {
            self.error(
                format!(
                    "Duplicate mesh id '{}' (first declared on line {})",
                    stmt.id, line
                ),
                stmt.line,
                stmt.col,
            );
        } else {
            self.mesh_ids.push((stmt.id.clone(), stmt.line));
        }
    }

    fn validate_connects(&mut self, program: &Program) {
        let all = collect_connects(&program.body);
        for c in all {
            if find_id(&self.channel_ids, &c.channel).is_none() {
                self.error(
                    format!("connect references undeclared channel '{}'", c.channel),
                    c.line,
                    c.col,
                );
            }
            if find_id(&self.node_ids, &c.node).is_none() {
                self.error(
                    format!("connect references undeclared node '{}'", c.node),
                    c.line,
                    c.col,
                );
            }
            if !VALID_TRANSPORTS.contains(&c.transport.as_str()) {
                self.error(
                    format!(
                        "Invalid transport '{}'; expected one of bluetooth | tcp | ws | ble5",
                        c.transport
                    ),
                    c.line,
                    c.col,
                );
            }
        }
    }
}

impl Default for Analyzer {
    fn default() -> Self {
        Self::new()
    }
}

fn find_id(ids: &[(String, usize)], id: &str) -> Option<usize> {
    ids.iter().find(|(k, _)| k == id).map(|(_, l)| *l)
}

/// Flatten a body (recursing into guard/cell blocks) into a flat list.
fn flatten_body(body: &[Statement]) -> Vec<&Statement> {
    let mut out: Vec<&Statement> = Vec::new();
    for s in body {
        out.push(s);
        match s {
            Statement::GuardStmt(g) => out.extend(flatten_body(&g.else_body)),
            Statement::CellBlock(c) => out.extend(flatten_body(&c.body)),
            _ => {}
        }
    }
    out
}

fn collect_connects(body: &[Statement]) -> Vec<&ConnectStmt> {
    let mut out: Vec<&ConnectStmt> = Vec::new();
    for s in body {
        match s {
            Statement::ConnectStmt(c) => out.push(c),
            Statement::GuardStmt(g) => out.extend(collect_connects(&g.else_body)),
            Statement::CellBlock(c) => out.extend(collect_connects(&c.body)),
            Statement::FnDecl(f) | Statement::PureFnDecl(f) => {
                out.extend(collect_connects(&f.body))
            }
            _ => {}
        }
    }
    out
}

/// Reimplements `/\b(await|console|output|fetch|Math\.random|Date\.now)\b|\.send\(|\.scan\(/`.
fn impure_test(raw: &str) -> bool {
    contains_word(raw, "await")
        || contains_word(raw, "console")
        || contains_word(raw, "output")
        || contains_word(raw, "fetch")
        || contains_word(raw, "Math.random")
        || contains_word(raw, "Date.now")
        || raw.contains(".send(")
        || raw.contains(".scan(")
}

/// Convenience wrapper: analyze a program in one call.
pub fn analyze(program: &Program) -> AnalysisResult {
    Analyzer::new().analyze(program)
}
